### PR TITLE
Implement logical exprs with multiple narrowings

### LINKED ISSUE #101: Implement logical expressions `&&` and `||` in the front-end
Specified in

https://ballerina.io/spec/lang/2021R1/#section_6.31

I don't think they need anything from the backend.

### PR DESCRIPTION
**Pull request by manuranga**
Fixes #101

### CODE REVIEW

**Comment by manuranga** [comment]
This doesn't fix #879. I'll send a separate pr for that.

File: compiler/modules/front/stmt.bal [hunk 1]
```diff
@@ -215,6 +216,28 @@ class StmtContext {
         return resolveSubsetTypeDesc(self.mod, self.functionDefn, td);
     }
 
+    function addNarrowings(bir:BasicBlock bb, Environment env, Narrowing[] narrowings, Position pos) {
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: Why is this a method?

File: compiler/modules/front/stmt.bal [hunk 2]
```diff
@@ -332,31 +354,31 @@ function codeGenStmt(StmtContext cx, bir:BasicBlock? curBlock, Environment env,
...
     else {
-        updateAssignments(env, effect);
-        addIntersectStmtNarrowings(narrowings, effect.narrowings);
+        updateAssignments(env.assignments, effect);
+        return combineNarrowings(narrowings, effect.narrowings, intersectNarrowing);
     }
 }
 
-function updateAssignments(Environment env, StmtEffect effect) {
+function updateAssignments(Assignment[] assignments, StmtEffect effect) {
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: This needs a longer name.

File: compiler/modules/front/stmt.bal [hunk 3]
```diff
@@ -332,31 +354,31 @@ function codeGenStmt(StmtContext cx, bir:BasicBlock? curBlock, Environment env,
     }
 }
 
-function applyEffect(Environment env, StmtNarrowing[] narrowings, StmtEffect effect) {
+function applyEffect(Environment env, Narrowing[] narrowings, StmtEffect effect) returns Narrowing[] {
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: This needs a longer name, which better conveys what it's doing.

File: compiler/modules/front/expr.bal [hunk 4]
```diff
@@ -703,9 +614,93 @@ function codeGenLogicalBinaryExpr(ExprContext cx, bir:BasicBlock bb, s:BinaryLog
...
+            i = i + 1;
+        }
+        if rnUsed != () {
+            rnsUsed[rnUsed] = true;
+        }
+        else {
+            pushNonNil(result, combine(ln, ()));
+        }
+    }
+    foreach var i in 0..<rnsLen {
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: Space before rnsLen

File: compiler/modules/front/expr.bal [hunk 5]
```diff
@@ -703,9 +614,93 @@ function codeGenLogicalBinaryExpr(ExprContext cx, bir:BasicBlock bb, s:BinaryLog
...
+        if !rnsUsed[i] {
+            pushNonNil(result, combine((), rns[i]));
+        }
+    }
+    return result;
+}
+
+function pushNonNil(Narrowing[] narrowings, Narrowing? n) {
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: Put narrowing in the name

File: compiler/modules/front/expr.bal [hunk 6]
```diff
@@ -703,9 +614,93 @@ function codeGenLogicalBinaryExpr(ExprContext cx, bir:BasicBlock bb, s:BinaryLog
...
+        };
+    }
+    else {
+        return (); // union with no narrowing is no narrowing
+    }
+}
+
+// May get removed with #ballerina-spec/1034
+// ln.binding.name == rn.binding.name if non-nil
+function unionNarrowing2(Narrowing? ln, Narrowing? rn) returns Narrowing? {
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: Name does not describe the difference from unionNarrowing.

**Review by jclark — COMMENTED** [review]
I've only done a cosmetic review.  We should go over this together in more detail. Overall looks nice.

File: compiler/modules/front/stmt.bal (hunk 1 above)
Comment (manuranga) [reply compiler/modules/front/stmt.bal]: It has to be called form expr.bal, I assumed expr.bal shouldn't call functions in stmt.bal (unless methods in context).

**Pull request merged**